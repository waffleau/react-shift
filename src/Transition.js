import React, {PureComponent, PropTypes} from 'react'

const EasingTypes = {
  easeIn: 'ease-in',
  easeInOut: 'ease-in-out',
  easeOut: 'ease-out',
  linear: 'linear'
}

const BaseProps = {
  active: PropTypes.bool.isRequired,
  children: PropTypes.any,
  delay: PropTypes.number,
  duration: PropTypes.number,
  easing: PropTypes.oneOf(Object.keys(EasingTypes)),
  style: PropTypes.object,
  timing: PropTypes.string
}

export class Transition extends PureComponent {
  static baseProps = BaseProps

  static defaultProps = {
    active: true,
    delay: 0,
    duration: 600,
    style: {}
  }

  static propTypes = {
    ...BaseProps,
    fromValue: PropTypes.object.isRequired,
    toValue: PropTypes.object.isRequired
  }

  getEasing() {
    const {easing, timing} = this.props
    return EasingTypes[easing] || timing || EasingTypes.linear
  }

  getProperty() {
    return Object.keys(this.props.fromValue)[0]
  }

  getStyle() {
    const {active, fromValue, style, toValue} = this.props

    return {
      transition: this.getTransition(),
      ...(active ? toValue : fromValue),
      ...style
    }
  }

  getTransition() {
    const {delay, duration} = this.props

    return `${this.getProperty()} ${duration}ms ${this.getEasing()} ${delay}ms`
  }

  render() {
    const {children} = this.props
    return (
      <div style={this.getStyle()}>
        {children}
      </div>
    )
  }
}

export default Transition
