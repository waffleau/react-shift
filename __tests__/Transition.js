import React from 'react'
import renderer from 'react-test-renderer'

import Transition from '../src/Transition'

test('can be inactive', () => {
  const tree = renderer.create(
    <Transition
      active={false}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can be active', () => {
  const tree = renderer.create(
    <Transition
      active
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom delay', () => {
  const tree = renderer.create(
    <Transition
      active
      delay={2222}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom entry delay', () => {
  const tree = renderer.create(
    <Transition
      active
      delayEnter={2222}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom exit delay', () => {
  const tree = renderer.create(
    <Transition
      active
      delayLeave={2222}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom duration', () => {
  const tree = renderer.create(
    <Transition
      active
      duration={9999}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom entry duration', () => {
  const tree = renderer.create(
    <Transition
      active
      durationEntry={9999}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom exit duration', () => {
  const tree = renderer.create(
    <Transition
      active
      durationLeave={9999}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom easing', () => {
  const tree = renderer.create(
    <Transition
      active
      easing="easeInOut"
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})
test('can have a custom easing', () => {
  const tree = renderer.create(
    <Transition
      active
      easing="easeInOut"
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('accepts an array of numbers describing easing', () => {
  const tree = renderer.create(
    <Transition
      active
      easing={[.1, .3, .7, .9]}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom entry easing', () => {
  const tree = renderer.create(
    <Transition
      active
      easingEntry="easeInOut"
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom exit easing', () => {
  const tree = renderer.create(
    <Transition
      active
      easingLeave="easeInOut"
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('can have a custom style', () => {
  const tree = renderer.create(
    <Transition
      active
      style={{ color: 'yellow' }}
      fromValue={{ backgroundColor: 'red' }}
      toValue={{ backgroundColor: 'blue' }}>
      {'Children'}
    </Transition>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})
