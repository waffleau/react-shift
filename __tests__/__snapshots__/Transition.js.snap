exports[`test can be active 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can be inactive 1`] = `
<div
  style={
    Object {
      "backgroundColor": "red",
      "transition": "backgroundColor 600ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom delay 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms linear 2222ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom duration 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 9999ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom easing 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms ease-in-out 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom entry delay 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms linear 2222ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom entry duration 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom entry easing 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom exit delay 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom exit duration 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom exit easing 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom style 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "color": "yellow",
      "transition": "backgroundColor 600ms linear 0ms",
    }
  }>
  Children
</div>
`;

exports[`test can have a custom timing 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms cubic-bezier(0.250, 0.250, 0.750, 0.750) 0ms",
    }
  }>
  Children
</div>
`;

exports[`test prefers easing over timing 1`] = `
<div
  style={
    Object {
      "backgroundColor": "blue",
      "transition": "backgroundColor 600ms ease-in-out 0ms",
    }
  }>
  Children
</div>
`;
